packages/muelu/src/Graph/BrickAggregation/MueLu_BrickAggregationFactory_def.hpp

```diff
@@ -480,6 +482,106 @@ namespace MueLu {
   }
 
 
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void BrickAggregationFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::BuildGraph(Level& currentLevel, const RCP<Matrix>& A) const {
+    // TODO: Currently only works w/ 1 DOF per node
+    double dirichletThreshold = 0.0;
+
+    if(bx_ > 1 && (nDim_ <= 1 || by_ > 1) && (nDim_ <=2 || bz_>1) ) {
+      FactoryMonitor m(*this, "Generating Graph (trivial)", currentLevel);
+      /*** Case 1: Use the matrix is the graph ***/
+      // Bricks are of non-trivial size in all active dimensions
+      RCP<GraphBase> graph = rcp(new Graph(A->getCrsGraph(), "graph of A"));
+      ArrayRCP<bool > boundaryNodes = Teuchos::arcp_const_cast<bool>(MueLu::Utilities<SC,LO,GO,NO>::DetectDirichletRows(*A, dirichletThreshold));
+      graph->SetBoundaryNodeMap(boundaryNodes);
+      
+      if (GetVerbLevel() & Statistics1) {
+        GO numLocalBoundaryNodes  = 0;
+        GO numGlobalBoundaryNodes = 0;
+        for (LO i = 0; i < boundaryNodes.size(); ++i)
+          if (boundaryNodes[i])
+            numLocalBoundaryNodes++;
+        RCP<const Teuchos::Comm<int> > comm = A->getRowMap()->getComm();
+        MueLu_sumAll(comm, numLocalBoundaryNodes, numGlobalBoundaryNodes);
+        GetOStream(Statistics1) << "Detected " << numGlobalBoundaryNodes << " Dirichlet nodes" << std::endl;
+      }      
+      Set(currentLevel, "DofsPerNode", 1);
+      Set(currentLevel, "Graph", graph);
+      Set(currentLevel, "Filtering",false);
+    }
+    else {
+      FactoryMonitor m(*this, "Generating Graph", currentLevel);
+      /*** Case 2: Dropping required ***/
+      // There is at least one active dimension in which we are not coarsening.  
+      // Those connections need to be dropped
+      bool drop_x = (bx_ == 1);
+      bool drop_y = (nDim_> 1 && by_ == 1);
+      bool drop_z = (nDim_> 2 && bz_ == 1);
+      
+      ArrayRCP<LO> rows   (A->getLocalNumRows()+1);
+      ArrayRCP<LO> columns(A->getLocalNumEntries());
+      
+      size_t N = A->getRowMap()->getLocalNumElements();
+
+      // FIXME: Do this on the host because indexing functions are host functions
```

Comment:
Does this imply data transfer to host is required?

Reply:
Well, BrickAggregation never did work on-device anyway, so my level of caring is low.